In [107]:
import pandas as pd
data = pd.read_csv("wnba-shots-2021.csv")

# ATTRIBUTES DESCRIPTION

In [108]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41497 entries, 0 to 41496
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   game_id                    41497 non-null  int64 
 1   game_play_number           41497 non-null  int64 
 2   desc                       41497 non-null  object
 3   shot_type                  41497 non-null  object
 4   made_shot                  41497 non-null  bool  
 5   shot_value                 41497 non-null  int64 
 6   coordinate_x               41497 non-null  int64 
 7   coordinate_y               41497 non-null  int64 
 8   shooting_team              41497 non-null  object
 9   home_team_name             41497 non-null  object
 10  away_team_name             41497 non-null  object
 11  home_score                 41497 non-null  int64 
 12  away_score                 41497 non-null  int64 
 13  qtr                        41497 non-null  int64 
 14  quarte

In [109]:
data.describe()

,game_id,game_play_number,shot_value,coordinate_x,coordinate_y,home_score,away_score,qtr,quarter_seconds_remaining,game_seconds_remaining
count,4.149700e+04,41497.000000,41497.000000,4.149700e+04,4.149700e+04,41497.000000,41497.000000,41497.000000,41497.000000,41497.000000
mean,4.013982e+08,191.296600,0.956672,-4.467604e+07,-4.467606e+07,42.536810,41.751090,2.519459,281.990770,1175.086657
std,1.917163e+04,114.060739,1.050284,8.716848e+07,8.716849e+07,25.396172,25.171965,1.138671,173.925427,701.455287
min,4.013916e+08,2.000000,0.000000,-2.147483e+08,-2.147484e+08,0.000000,0.000000,1.000000,0.000000,0.000000
25%,4.013917e+08,92.000000,0.000000,5.000000e+00,0.000000e+00,21.000000,20.000000,2.000000,128.000000,572.000000
50%,4.013918e+08,191.000000,1.000000,2.300000e+01,3.000000e+00,42.000000,41.000000,3.000000,279.000000,1180.000000
75%,4.013918e+08,287.000000,2.000000,2.700000e+01,1.400000e+01,63.000000,61.000000,4.000000,433.000000,1790.000000
max,4.014560e+08,533.000000,3.000000,5.000000e+01,8.100000e+01,134.000000,116.000000,6.000000,600.000000,2395.000000


# NA ANALYSIS

In [110]:
data.isnull().sum()

game_id                      0
game_play_number             0
desc                         0
shot_type                    0
made_shot                    0
shot_value                   0
coordinate_x                 0
coordinate_y                 0
shooting_team                0
home_team_name               0
away_team_name               0
home_score                   0
away_score                   0
qtr                          0
quarter_seconds_remaining    0
game_seconds_remaining       0
dtype: int64

# SHOT TYPE

In [111]:
len(list(data['shot_type'].unique())), data['shot_type'].value_counts().sort_values()

(58,
 shot_type
 Driving Dunk Shot                         1
 Alley Oop Dunk Shot                       1
 No Shot (Default Shot)                    1
 Running Dunk Shot                         1
 Free Throw - Flagrant 1 of 1              6
 Step Back Bank Jump Shot                  7
 Hook Driving Bank                        10
 Running Alley Oop Layup Shot             13
 Turnaround Fadeaway Bank Jump Shot       13
 Hook Shot Bank                           13
 Free Throw - Clear Path 1 of 2           14
 Free Throw - Clear Path 2 of 2           14
 Free Throw - Flagrant 1 of 3             15
 Free Throw - Flagrant 2 of 3             15
 Free Throw - Flagrant 3 of 3             15
 Fade Away Bank Jump Shot                 16
 Free Throw - Flagrant 1 of 2             28
 Free Throw - Flagrant 2 of 2             28
 Hook Turnaround Bank                     29
 Finger Roll Layup                        30
 Layup Running Reverse                    43
 Cutting Finger Roll Layup Shot        

There are too many shot types. We group them into 7 categories:

In [112]:
shot_type_to_group = {
    'Alley Oop Dunk Shot': 'Dunk',
    'Driving Dunk Shot': 'Dunk',
    'Running Dunk Shot': 'Dunk',
    'Alley Oop Layup Shot': 'Layup',
    'Running Alley Oop Layup Shot': 'Layup',
    'Cutting Layup Shot': 'Layup',
    'Cutting Finger Roll Layup Shot': 'Layup',
    'Driving Layup Shot': 'Layup',
    'Driving Finger Roll Layup': 'Layup',
    'Driving Floating Jump Shot': 'Layup',
    'Driving Floating Bank Jump Shot': 'Layup',
    'Layup Driving Reverse': 'Layup',
    'Running Layup Shot': 'Layup',
    'Running Finger Roll Layup': 'Layup',
    'Layup Running Reverse': 'Layup',
    'Finger Roll Layup': 'Layup',
    'Layup Shot': 'Layup',
    'Layup Shot Putback': 'Layup',
    'Reverse Layup Shot': 'Layup',
    'Jump Shot': 'Jump Shot',
    'Floating Jump Shot': 'Jump Shot',
    'Pullup Jump Shot': 'Jump Shot',
    'Running Jump Shot': 'Jump Shot',
    'Running Pullup Jump Shot': 'Jump Shot',
    'Jump Shot Bank': 'Jump Shot',
    'Driving Jump Shot Bank': 'Jump Shot',
    'Fade Away Bank Jump Shot': 'Jump Shot',
    'Fade Away Jump Shot': 'Jump Shot',
    'Pullup Bank Jump Shot': 'Jump Shot',
    'Step Back Bank Jump Shot': 'Jump Shot',
    'Turnaround Bank Jump Shot': 'Jump Shot',
    'Step Back Jump Shot': 'Jump Shot',
    'Turnaround Fade Away Jump Shot': 'Jump Shot',
    'Turnaround Fadeaway Bank Jump Shot': 'Jump Shot',
    'Turnaround Jump Shot': 'Jump Shot',
    'Driving Hook Shot': 'Hook Shot',
    'Hook Shot': 'Hook Shot',
    'Hook Shot Bank': 'Hook Shot',
    'Hook Turnaround Bank': 'Hook Shot',
    'Turnaround Hook Shot': 'Hook Shot',
    'Free Throw - 1 of 1': 'Free Throw',
    'Free Throw - 1 of 2': 'Free Throw',
    'Free Throw - 1 of 3': 'Free Throw',
    'Free Throw - 2 of 2': 'Free Throw',
    'Free Throw - 2 of 3': 'Free Throw',
    'Free Throw - 3 of 3': 'Free Throw',
    'Free Throw - Clear Path 1 of 2': 'Free Throw',
    'Free Throw - Clear Path 2 of 2': 'Free Throw',
    'Free Throw - Flagrant 1 of 1': 'Free Throw',
    'Free Throw - Flagrant 1 of 2': 'Free Throw',
    'Free Throw - Flagrant 1 of 3': 'Free Throw',
    'Free Throw - Flagrant 2 of 2': 'Free Throw',
    'Free Throw - Flagrant 2 of 3': 'Free Throw',
    'Free Throw - Flagrant 3 of 3': 'Free Throw',
    'Free Throw - Technical': 'Free Throw',
    'Tip Shot': 'Tip Shot',
    'No Shot (Default Shot)': 'No Shot'
}

# Apply the mapping to create the 'grouped_shot_type' column
data['grouped_shot_type'] = data['shot_type'].map(shot_type_to_group)

data['grouped_shot_type'].value_counts()

grouped_shot_type
Jump Shot     20063
Layup         11973
Free Throw     8633
Hook Shot       665
Tip Shot        149
Dunk              3
No Shot           1
Name: count, dtype: int64

We could eliminate the 'Dunk' and 'No Shot' entries

# TEAM

In [113]:
data['shooting_team'].unique()

array(['Indiana', 'Washington', 'Los Angeles', 'Chicago', 'Phoenix',
       'Las Vegas', 'Minnesota', 'Seattle', 'Connecticut', 'New York',
       'Dallas', 'Atlanta', 'Team Wilson', 'Team Stewart'], dtype=object)

# PLAYER

In [114]:
data['desc'].head()

0    Shatori Walker-Kimbrough blocks Destanni Hende...
1              Elena Delle Donne misses two point shot
2    Tiffany Mitchell makes 4-foot layup (Destanni ...
3                    Natasha Cloud makes driving layup
4    Tiffany Mitchell makes 26-foot three point jum...
Name: desc, dtype: object

In [115]:
import re
def find_non_capitalized(text):
    word_list = text.split(" ")
    for word in word_list:
        if not word[0].isupper():
            return word

# Apply the function to the 'desc' column
data['first_non_capitalized'] = data['desc'].apply(find_non_capitalized)

In [116]:
data['first_non_capitalized'].value_counts()

first_non_capitalized
makes     21405
misses    18277
blocks     1815
Name: count, dtype: int64

In [117]:
data[data['first_non_capitalized']=='blocks']['desc'].unique()

array(["Shatori Walker-Kimbrough blocks Destanni Henderson 's 15-foot jumper",
       "Natasha Cloud blocks Victoria Vivians 's 20-foot pullup jump shot",
       "Emily Engstler blocks Myisha Hines-Allen 's 5-foot layup", ...,
       "A'ja Wilson blocks Courtney Williams 's 4-foot layup",
       "Alyssa Thomas blocks Kelsey Plum 's 7-foot driving layup",
       "Chelsea Gray blocks Alyssa Thomas 's 3-foot two point shot"],
      dtype=object)

In [118]:
def find_blocking_player(row):
    if row['first_non_capitalized'] == 'blocks':
        word_list = row['desc'].split(" ")
        player_name = ""
        for word in word_list:
            if word[0].isupper():
                player_name = player_name + word + " "
            else:
                break
        player_name = player_name[:-1]
        return player_name
data['blocking_player'] = data.apply(lambda row: find_blocking_player(row),axis =1)

In [119]:
def find_shooting_player(row):
    word_list = row['desc'].split(" ")
    if row['first_non_capitalized'] != 'blocks':
        player_name = ""
        for word in word_list:
            if word[0].isupper():
                player_name = player_name + word + " "
            else:
                break
        player_name = player_name[:-1]
    else:
        player_name = ""
        for word in word_list[word_list.index('blocks')+1:]:
            if word[0].isupper():
                player_name = player_name + word + " "
            else:
                break
        player_name = player_name[:-1].replace("'s","")
    return player_name

In [120]:
prova = "Joel Sasot Ruiz blocks Cristiano Ronaldo a 6-foot shit"
prova.split().index('blocks')


3

In [121]:
data['shooting_player'] = data.apply(lambda row: find_shooting_player(row),axis =1)

# COORDINATES STUDY

In [122]:
data[['coordinate_x','coordinate_y']].describe()


,coordinate_x,coordinate_y
count,4.149700e+04,4.149700e+04
mean,-4.467604e+07,-4.467606e+07
std,8.716848e+07,8.716849e+07
min,-2.147483e+08,-2.147484e+08
25%,5.000000e+00,0.000000e+00
50%,2.300000e+01,3.000000e+00
75%,2.700000e+01,1.400000e+01
max,5.000000e+01,8.100000e+01


## COORDINATE Y CORRECTION
The regulation states that the hoop is placed 63'' (5.25 feet) from the base line. In this document the hoops location seems to be at 0. This must be corrected in order to have all 'coordinate_y' values positive (and facilitate its analysis).

In [123]:
data['coordinate_y'] = data['coordinate_y']+5

## FREE SHOT CORRECTION

In [124]:
data[data['grouped_shot_type']=='Free Throw']['coordinate_x'].value_counts()

coordinate_x
-214748340    8633
Name: count, dtype: int64

In [125]:
def free_throw_coordinate_x_correction(row):
    if row['grouped_shot_type']=='Free Throw':
        return 25
    return row['coordinate_x']

data['coordinate_x'] = data.apply(lambda row: free_throw_coordinate_x_correction(row),axis =1)

In [126]:
data[data['grouped_shot_type']=='Free Throw']['coordinate_x'].value_counts()

coordinate_x
25    8633
Name: count, dtype: int64

In [127]:
data[data['grouped_shot_type']=='Free Throw']['coordinate_y'].value_counts()

coordinate_y
-214748360    8633
Name: count, dtype: int64

In [128]:
def free_throw_coordinate_y_correction(row):
    if row['grouped_shot_type']=='Free Throw':
        return 15
    return row['coordinate_y']
data['coordinate_y'] = data.apply(lambda row: free_throw_coordinate_y_correction(row),axis =1)


In [129]:
data[data['grouped_shot_type']=='Free Throw']['coordinate_y'].value_counts()

coordinate_y
15    8633
Name: count, dtype: int64

In [130]:
data[['coordinate_x','coordinate_y']].describe()


,coordinate_x,coordinate_y
count,41497.000000,41497.000000
mean,24.837169,14.520399
std,9.465879,7.812525
min,0.000000,0.000000
25%,22.000000,7.000000
50%,25.000000,15.000000
75%,27.000000,19.000000
max,50.000000,86.000000


## OWN FIELD SHOTS

The range of the coordinate_y values seems to take into the length of the complete court. However, if the amount of shots taken from a team's own field is small enough, they can be overseen.

In [131]:
data[data['coordinate_y']>47].shape,data.shape

((60, 20), (41497, 20))

There are 60 out of 41497 shots which are attempted from a team's own field (a little over 0.001%). Let us check if there might be a team that has made this kind of shots their signature:

In [132]:
data[data['coordinate_y']>47]['shooting_team'].value_counts()

shooting_team
Washington     8
Chicago        8
Indiana        7
Atlanta        6
Dallas         6
New York       6
Connecticut    5
Los Angeles    4
Phoenix        4
Las Vegas      2
Seattle        2
Team Wilson    1
Minnesota      1
Name: count, dtype: int64

No team stands out, so the 60 shots are deleted.

In [133]:
data = data[data['coordinate_y']<=47]

In [134]:
data[['coordinate_x','coordinate_y']].describe()


,coordinate_x,coordinate_y
count,41437.000000,41437.000000
mean,24.834472,14.459613
std,9.462392,7.645660
min,0.000000,0.000000
25%,22.000000,7.000000
50%,25.000000,15.000000
75%,27.000000,19.000000
max,50.000000,47.000000


# Data per team

In [135]:
data['shooting_team'].value_counts()

shooting_team
Connecticut     4227
Las Vegas       4144
Chicago         3817
Seattle         3535
Dallas          3434
Phoenix         3240
Washington      3199
Indiana         3188
New York        3182
Minnesota       3112
Atlanta         3081
Los Angeles     3062
Team Stewart     109
Team Wilson      107
Name: count, dtype: int64

Team Stewart and Team Wilson have a different order of magnitude of data, so they are not included.

In [136]:
data = data[~data['shooting_team'].isin(['Team Stewart','Team Wilson'])]

# **Shot value**

In [137]:
data.columns

Index(['game_id', 'game_play_number', 'desc', 'shot_type', 'made_shot',
       'shot_value', 'coordinate_x', 'coordinate_y', 'shooting_team',
       'home_team_name', 'away_team_name', 'home_score', 'away_score', 'qtr',
       'quarter_seconds_remaining', 'game_seconds_remaining',
       'grouped_shot_type', 'first_non_capitalized', 'blocking_player',
       'shooting_player'],
      dtype='object')

The shot value is always 0 points in case of a miss. It could prove useful to quickly know when a shot would have been worth 1,2 or 3 points in case of having been made. There are bound to be some mistaken imputations due to the integer format of the coordinates and shot characteristics (if the feet of the shooter leave the ground outside the line...)

In [138]:
import numpy as np

def shot_value(x, y):
    """
    Determine the shot value based on coordinates (x, y).
    
    Parameters:
    x (int): Horizontal position (0 to 50)
    y (int): Vertical position (0 to 47)
    
    Returns:
    int: 2 or 3 depending on shot zone
    """
    # Hoop coordinates
    hoop_x, hoop_y = 25, 5
    
    # Calculate distance from hoop
    distance = np.sqrt((x - hoop_x) ** 2 + (y - hoop_y) ** 2)
    
    # WNBA 3-point range thresholds
    three_point_distance_center = 22.15 # 3-point arc in the center
    three_point_distance_sides = 22.0    # Shorter arc on the sides
    
    # Check if within 3-point range
    if y > 14 and distance > three_point_distance_center:
        return 3
    elif y <= 14 and distance > three_point_distance_sides:
        return 3
    else:
        return 2



In [139]:
def impute_max_points(row):
    if row['grouped_shot_type'] == 'Free Throw':
        return 1
    if row['shot_value']!=0:
        return row['shot_value']
    return shot_value(row['coordinate_x'], row['coordinate_y'])

In [140]:
data['max_value'] = data.apply(impute_max_points,axis=1)

In [141]:
data.describe()

,game_id,game_play_number,shot_value,coordinate_x,coordinate_y,home_score,away_score,qtr,quarter_seconds_remaining,game_seconds_remaining,max_value
count,4.122100e+04,41221.000000,41221.000000,41221.000000,41221.000000,41221.000000,41221.000000,41221.000000,41221.000000,41221.000000,41221.000000
mean,4.013979e+08,191.116251,0.956818,24.835302,14.442881,42.426409,41.687368,2.519856,282.403459,1175.293006,2.048883
std,1.875501e+04,114.014928,1.049308,9.455346,7.623263,25.274025,25.121758,1.138994,173.692967,701.641649,0.681610
min,4.013916e+08,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,1.000000
25%,4.013917e+08,92.000000,0.000000,22.000000,7.000000,21.000000,20.000000,2.000000,129.000000,571.000000,2.000000
50%,4.013918e+08,190.000000,1.000000,25.000000,15.000000,42.000000,41.000000,3.000000,279.000000,1180.000000,2.000000
75%,4.013918e+08,287.000000,2.000000,27.000000,19.000000,63.000000,61.000000,4.000000,434.000000,1790.000000,3.000000
max,4.014557e+08,533.000000,3.000000,50.000000,47.000000,118.000000,116.000000,6.000000,600.000000,2395.000000,3.000000


# **Saving the clean data**

In [142]:
data.to_csv('clean_data.csv',index = False)